### PR TITLE
Proposal: New Makefile standard template

### PR DESCRIPTION
**Pull request by jaekwon**
Lets use these conventions for Makefiles.  I've got a similar one for tendermint/tendermint.

Note the order for `all:` which gets run when you type `make`.

> check build test install metalinter

Check checks stuff without making side effects.  If there are tools that need to be installed, it yells at the user.  The user then has to run `make get_tools` manually, then type `make` again.

The build step quickly ensures that the project can actually compile.  Protoc and other file generations happen there as well.  Then you run the tests.  The tests may rely on the build step so this separation is important.

Installation should happen after tests pass.

The metalinter comes last.  It takes forever to run and it's annoying.  I don't mind if it hangs around on my terminal after I'm done though, so there it goes.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -21,60 +59,41 @@ install_protoc:
...
 test:
-	@ find . -path ./vendor -prune -o -name "*.sock" -exec rm {} \;
-	@ echo "==> Running go test"
-	@ go test $(PACKAGES)
+	@find . -path ./vendor -prune -o -name "*.sock" -exec rm {} \;
+	@echo "==> Running go test"
+	@go test $(PACKAGES)
 
 test_race:
-	@ find . -path ./vendor -prune -o -name "*.sock" -exec rm {} \;
```
Comment (odeke-em) [inline Makefile]: It's odd that we had a space between `@` and the `command`, thanks for fixing this!

File: Makefile [hunk 2]
```diff
@@ -21,60 +59,41 @@ install_protoc:
...
-	## On "error while loading shared libraries: libprotobuf.so.14: cannot open shared object file: No such file or directory"
-	##   ldconfig (may require sudo)
-	## https://stackoverflow.com/a/25518702
-	protoc $(INCLUDE) --gogo_out=plugins=grpc:. types/*.proto
-	@ echo "--> adding nolint declarations to protobuf generated files"
-	@ awk '/package types/ { print "//nolint: gas"; print; next }1' types/types.pb.go > types/types.pb.go.new
-	@ mv types/types.pb.go.new types/types.pb.go
+update_tools:
+	@echo "--> Updating tools"
+	@go get -u $(GOTOOLS)
```
Comment (odeke-em) [inline Makefile]: Clean! Me likey.

File: Makefile [hunk 3]
```diff
@@ -1,18 +1,56 @@
...
+	@echo "Found tools: $(foreach tool,$(GOTOOLS_CHECK),\
+        $(if $(shell which $(tool)),$(tool),$(error "No $(tool) in PATH")))"
+
+get_tools:
+	@echo "--> Installing tools"
+	go get -u -v $(GOTOOLS)
+	@gometalinter.v2 --install
+
+get_protoc:
+	@# https://github.com/google/protobuf/releases
```
Comment (odeke-em) [inline Makefile]: Why the commented directives?

Comment (jaekwon) [reply Makefile]: Otherwise it prints onto the console, as if you typed it into bash.

File: Makefile [hunk 4]
```diff
@@ -1,18 +1,56 @@
...
+	github.com/gogo/protobuf/gogoproto
+GOTOOLS_CHECK = gox glide gometalinter.v2 protoc protoc-gen-gogo
+PACKAGES=$(shell go list ./... | grep -v '/vendor/')
 INCLUDE = -I=. -I=${GOPATH}/src -I=${GOPATH}/src/github.com/gogo/protobuf/protobuf
 
-all: protoc install test metalinter
+all: check build test install metalinter
 
-PACKAGES=$(shell go list ./... | grep -v '/vendor/')
+check: check_tools get_vendor_deps
```
Comment (ebuchman) [inline Makefile]: should get_vendor_deps be part of this ? If `check_tools` says `glide` is not installed, `get_vendor_deps` will fail ...

**Comment by ebuchman** [comment]
need to update circle.yml to ensure tools are installed first.

File: Makefile (hunk 4 above)
Comment (jaekwon) [reply Makefile]: if glide is not installed, check_tools will cause Makefile to exit.

File: Makefile [hunk 5]
```diff
@@ -1,18 +1,56 @@
...
+	github.com/gogo/protobuf/gogoproto
+GOTOOLS_CHECK = gox glide gometalinter.v2 protoc protoc-gen-gogo
+PACKAGES=$(shell go list ./... | grep -v '/vendor/')
 INCLUDE = -I=. -I=${GOPATH}/src -I=${GOPATH}/src/github.com/gogo/protobuf/protobuf
 
-all: protoc install test metalinter
+all: check protoc build test install metalinter
 
-PACKAGES=$(shell go list ./... | grep -v '/vendor/')
+check: check_tools get_vendor_deps
```
Comment (ebuchman) [inline Makefile]: Opening comment says `Check checks stuff without making side effects.`. So this probably shouldn't have `get_vendor_deps`

Comment (jaekwon) [reply Makefile]: Yeah.  We can do:
```
all: check get_vendor_deps protoc build test install metalinter
check: check_tools
```

**Comment by jaekwon** [comment]
`make` shouldn’t be installing tools, they should be pre-requisites for calling `make`.  `make install` should not install tools either, generally you call `make` and then `make install`.  In our case we also install upon `make`.  Lets make it a standard to require `make get_tools` manually.

**Comment by ebuchman** [comment]
@Greg-Szabo probably best if the images we use are baked with our `make get_tools` so we don't have disparities. 

Of course we still need a way to version these tools.

@jaekwon what do you think? Should we keep a json of tools and their git commits and install from source?

**Pull request merged**